capsules/src/usb/descriptors.rs

```diff
@@ -370,6 +375,183 @@ impl Descriptor for DeviceDescriptor {
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(), Cell::default(), Cell::default(),
+            Cell::default(), Cell::default(), Cell::default(),
+        ],
+        len: 0,
+    };
+
+    // Setup certain descriptor fields since now we know the tree of
+    // descriptors.
+
+    // Configuration Descriptor. We assume there is only one configuration
+    // descriptor, since this is very common for most USB devices.
+    configuration_descriptor.num_interfaces = interface_descriptor.len() as u8;
+
+    // Calculate the length of all dependent descriptors.
+    // TODO should we be erroring here if len > 128? Otherwise we'll probably
+    // buffer overrun and panic.
```

Comment:
This should definitely be handled explicitly, e.g. `panic!("Configuration descriptor too large")`. This shouldn't be a problem given that all of the configuration is defined at compile time, so even if this panic happens at runtime, it's reliably triggered depending on the compile-time descriptors.

This is also IMO a reason why these descriptors should rather be serialized on-demand following get-descriptor requests (which then doesn't impose a limit on the total buffer size), rather than eagerly serialized in a temporary buffer here.

Reply:
What _is_ the best way to handle this? We don't want `panic!` in capsules generally, but if this overruns we do want a fast, obvious error.

Just leaving a note for now, is honestly probably sufficient for the foreseeable future (at least until a USB capsule cleanup and rewrite) since 128 bytes fits CDC, OpenSK stuff, and also would fit USB Mass Storage or additional HID stuff if we want to expand to that in the future.

Reply:
Unfortunately, I think even serializing on demand doesn't actually fix this simply. You would just end up with structs that, when serialized sometime at runtime, end up either overrunning the buffer or more likely getting truncated. You could handle this at runtime, but doing so would require you to account for partial progress in serializing some struct or another and picking up serializing part-way through that struct for the next packet. (You can't just stop at the last struct that fits because I believe you have to send the full 64 byte packet in order to fragment.) At least with the descriptors serialized at initialization time, the error occurs immediately.

Reply:
> What _is_ the best way to handle this? We don't want `panic!` in capsules generally, but if this overruns we do want a fast, obvious error.

I don't see any `unsafe` in this file. So if the buffer overruns, the current behavior is that it *will panic* at one of the array indexing operations (e.g. within a `write_to` where the target buffer is smaller than the source). So the current code would yield a panic (which we want to avoid apparently), a fast error (in this function), but not so obvious (array indexing out of bounds somewhere).

What I suggest is an explicit panic with a clear message here, which improves over the current state by making the error obvious.

To avoid a panic altogether, the best would probably be to put all of this code in a `const fn` so that the contents of the descriptor buffer is pre-computed at compile time, and that any failure would prevent it from compiling. However, I'm not sure what the current status of `const fn` is in terms of supporting things like loops (although I remember there has been some improvements lately).